### ISSUE TITLE
Feedback for parameter values

### ISSUE THREAD

**Issue by sstaub** [bitfocus/companion-module-etc-eos#16]
I experimented with the encoders from Stream Deck + to set parameter values which works fine.
But a feedback will welcome to display the parameter values.

**Comment by n9yty** [comment]
Is it able to act as encoders for pan/tilt/color/etc ?

**Comment by n9yty** [comment]
I am working on capturing and setting variables for a variety of parameters. See recent betas in v3. Right now I am capturing mostly wheel/encoder parameters, but I have worked with MBresin's ETC-NOMAD-SD configuration and have added feedback on the encoder display of the StreamDeck+. Perhaps he will release it when he feels it is working properly and updated to take advantage of it. I have tweaked it (and changed the Companion module to set the variables) but I don't think I have the right to redistribute it.

**Comment by sstaub** [comment]
Parameter feedback is very tricky, I'm working on an implementation for Arduino.
There are two possibilities
— subscribing parameters to get the value for the parameter name, this must done directly after connection is established
— processing of the wheel numbers, to get parameters based on the category. The wheel list change depending on the parameter footprint of the selected fixtures. The parameters have no fixed wheel number.

**Comment by n9yty** [comment]
I am capturing them by viewing the label returned by the wheel OSC messages and assigning to variables for known parameters. It is working very well in the latest companion v3 beta builds. I just pushed some additional changes through that fix some little quirks I let creep in.

**Comment by sstaub** [comment]
I will give a try. I see that animation, gobo, and prism wheels are missing.

**Comment by n9yty** [comment]
Public call to document missing parameters. ;)

If you can spell out what you are missing, I can add the variables. I have limited EOS experience at the point, nothing with any of those effects, so I don't know what I don't know about what is needed.

**Comment by sstaub** [comment]
I will add a list the next days. To see what wheel/parameter are sending simple use the diagnostic tab 99 and set OSC send logging.

**Comment by n9yty** [comment]
Yes, but I don't know which fixtures to try as they all send different wheels as you know, so hard to know what all to look at to capture all that may be useful. Feedback is more important.

**Comment by sstaub** [comment]
Missing, taken from some popular fixtures from ETC, HighEnd, Martin Robe:
Cooling Fan
Dim Curve
Global MSpeed
Lt Diffusion
Med Diffusion
Color Mix Mode
Edge Mode
Edge Adjust
Effect Macros
Form MSpeed
Tint
Color Temperature
Color Mix Mode 2

I think there are many other parameter (about 2000) but not all are useful.
As a hint, the parameter names from the wheel are normally for displaying.
If you want sending this parameter with /eos/wheel/<parameter> you must replace the slash "/" in the parameter name with a backslash "\"

**Comment by n9yty** [comment]
Thank your for that list. If  possible, do you have the exact fixture names I could test against in EOS? For example, when doing Gobo select I found that you could not send the update as a wheel change with a 1 or -1 parameter, you had to send the gobo number in a param update. I suppose that is only important for whoever implements the button in Companion, not for the module which is simply reporting the value in a variable. I will add these to the list of variables that are populated.

**Comment by sstaub** [comment]
Selection of gobo and others did not really work. This is by system.
For fixtures try Martin Viper, HighEnd Solarframe Theatre, HighEnd Lonestar, Robe T2 Profile,Etc Sourcfour LED Series 3

**Comment by n9yty** [comment]
Thank you. I did come up with logic to make rotating with the know select Gobos, so perhaps I can test the others.

**Comment by sstaub** [comment]
Wow, this is was we expected, needed, wished, thank you. On the first look I was very impressed.
I think that there are some thing which could better. 
1. If you deselect the channel, the values stays. If you receive a category with value 0 for a wheel, there is no more value and parameter for it.
2. Value resolution should 3 digits after point, in the moment there are also 2. e.g. Pan 2.00 vs 2.001, this important for fine adjustment
3. Course, Fine, Acceleration 
Fine doesn't work as expected you must send `/eos/wheel/fine/<Parameter>` for the correct behavior of the parameter. Course gives the MSB and fine the LSB of of a 16bit Parameter.
Also you should "Course" rename in "Acc" (Acceleration) because "course" is the regular mode by design.
4. Parameter "Frame Assembly" is missing in Shutters 
5. Brightness setting, please some more steps instead of two.

Two general things
1. As an advice with v3.2.x it is possible to use the loopback address 127.0.0.1 instead of a real active network connection, it is only necessary that a device is recognized neither if it is online or not.
2. Use TCP 1.1 Slip on Port 3037, this is recommended with v3.x for third party applications. Maybe this should be an backward compatibility option for users which use a legacy versions 2.9.x

**Comment by n9yty** [comment]
@sstaub I tried to capture when fixture changes and reset values. I have not found a reliable way (based on watching OSC data) to capture it in every case. I will continue to look for ways to improve this.

I can adjust value precision, and will push that through in an upcoming update. I will wait for more feedback to consolidate changes.

I will look into using /fine/ instead of /course/, thank you for pointing that out. Hopefully it is possible to do given the constraints of Companion. I'm impressed I have as much working on one page as I do. :)

For Frame Assembly, the parameters all come from whatever the fixture returns. Nothing is hard-coded. This is for all the dynamic encoders. Please advise which fixture you are missing that on, and if possible even turn on OSC debugging in EOS and see if it is sending that value.

I do have the unique parameter trying to capture it unless I have the returned name wrong:

'frame assembly':'enc_frame_assembly'

I wish Brightness would respond to more changes, but it only takes a fixed value, so I have used the options of left and right rotation and pressing. I cannot even manipulate a variable and set it because it doesn't take a parameter, just a slider that is fixed in the definition. That would be a feature request to Bitfocus in the Companion software itself.

I do not, at this time, think coding to just 3.2.x is best practice. That could change in the future, and besides, you could be running this on a RaspberryPi or something else remote from the actual EOS software. It is configurable, though, so nothing is locked in. In testing I develop on my Mac but have the software running on a miniPC so 127.0.0.1 won't work in that case anyway. Feel free to adjust as best first your use case.

On TCP 1.1/etc - I know you have asked for this before in other requests, at this time it is not something I am willing to jump in on, but it is open source. :)

**Comment by n9yty** [comment]
Try this config file to fix the coarse/fine and shutter frame assembly. The precision is fixed in the module code which is not pushed up yet. 

[sample_config_streamdeckplus.companionconfig.zip](https://github.com/bitfocus/companion-module-etc-eos/files/11709218/sample_config_streamdeckplus.companionconfig.zip)

**Comment by n9yty** [comment]
@sstaub There may be other issues, remember, I don't actually *use* this as I don't do lighting work. I am thankful for your feedback to help make it better.

This should fix the missing "Frame Assembly" in Shutters (and any other additional parameters in the Shutters Category, it was counting that one set of page numbers wrong).

It should also fix the coarse/fine issue, please test.

This does not fix the precision (.xxx versus .xx) because that is in the module code that is not yet pushed for updates, but it does work in my testing.

I will look at renaming Course to Accel, but that is only visual so for now this could be usable for testing functionality.

[sample_config_streamdeckplus.companionconfig.zip](https://github.com/bitfocus/companion-module-etc-eos/files/11709283/sample_config_streamdeckplus.companionconfig.zip)

**Comment by sstaub** [comment]
@n9yty I know about the dynamic wheel proceeding, I'm working for a microcontroller Arduino  C/C++ solution and it is very hard to realize because you don't know when all wheels are completely send.
Maybe a very short timeout (small enough to collect all wheels) will a workaround for updating all stuff inside. 
I made a request to ETC that they should send the number of wheels when the selection changes but I don't know if it happens in future.
For SLIP / TCP 1.1, I think it is supported by the OSC library you use.
Is there meanwhile a way to send data with arguments/values?, see issue #17 

I wish I have more knowledge about .js, I tried to work on a fork of the library without luck (using macOS Ventura)
